src/NServiceBus.Transport.SQS/Renewal.cs

```diff
@@ -0,0 +1,163 @@
+                        ex);
+                }
+
+                // Signaling the message receipt handle is invalid so that other operations relaying on the token owned
+                // by this token source can be cancelled.
+                await messageVisibilityLostCancellationTokenSource.CancelAsync()
+                    .ConfigureAwait(false);
+                return Result.Failed;
+            }
+            catch (AmazonSQSException ex) when (ex.IsCausedByMessageVisibilityExpiry())
+            {
+                // This is debug level because we expect this to happen when the message is picked up by a competing consumer
+                // or when the message successfully completed processing and the message is deleted
+                if (Logger.IsDebugEnabled)
+                {
+                    Logger.Debug(
+                        $"Renewing the message visibility timeout failed because the message with native ID {receivedMessage.MessageId} has already been picked up by a competing consumer.",
+                        ex);
+                }
+
+                // Signaling the message receipt handle is invalid so that other operations relaying on the token owned
+                // by this token source can be cancelled.
+                await messageVisibilityLostCancellationTokenSource.CancelAsync()
+                    .ConfigureAwait(false);
+                return Result.Failed;
+            }
+            catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
+            {
+                LogStopped(receivedMessage, visibilityExpiresOn);
+                return Result.Stopped;
+            }
+            catch (Exception ex)
+            {
+                if (Logger.IsWarnEnabled)
+                {
+                    Logger.Warn($"Renewing the message visibility timeout failed for the message with native ID '{receivedMessage.MessageId}'.", ex);
+                }
+
+                return Result.Failed;
+            }
+        }
+
+        return Result.Stopped;
+    }
+
+    static void LogStopped(Message receivedMessage, DateTimeOffset visibilityExpiresOn)
+    {
+        if (Logger.IsDebugEnabled)
+        {
+            Logger.DebugFormat("The message visibility timeout renewal for message with native ID '{0}' was stopped. The message visibility might expire at '{1}' UTC", receivedMessage.MessageId, visibilityExpiresOn);
+        }
+    }
+
+    public enum Result
+    {
+        Failed,
+        Stopped
+    }
+
+    public static TimeSpan CalculateRenewalTime(DateTimeOffset visibilityTimeExpiresOn,
+        TimeProvider timeProvider = null)
+    {
+        timeProvider ??= TimeProvider.System;
+        var remainingTime = visibilityTimeExpiresOn - timeProvider.GetUtcNow();
+
+        if (remainingTime < TimeSpan.FromMilliseconds(400))
+        {
+            return TimeSpan.Zero;
+        }
```

Comment:
Shall we document what's unique about 400ms?

Reply:
I also guess that `return TimeSpan.Zero;` means renew immediately, right?

Reply:
Yes, that's used then in Task.Delay, which then won't delay basically